Add Utils class with helper methods for listing block template locations

lukecarbis opened the pull request:
Fix for a bug affecting the template location display when the template is in one of the new template locations.

Bug description: https://d.pr/v/iPUk0q

lukecarbis committed:
Add helper function for listing block template locations

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -398,27 +394,20 @@ public function block_template( $name, $type = 'block' ) {
 
 		$types         = (array) $type;
 		$located       = '';
 		$template_file = '';
```
It looks like this assignment of `$template_file` might not be needed anymore.

lukecarbis replied:
Fixed in 3fc23ef

kienstra replied:
Thanks, that looks good.

kienstra commented on php/helpers.php:
```diff
@@ -227,6 +227,22 @@ function block_lab_locate_template( $template_names, $path = '', $single = true
+ *
+ * @return array
+ */
+function block_lab_template_locations( $name, $type = 'block' ) {
```
With the current architecture, this `helpers.php` file is probably the best place for this function.

But in the future, it might be nice to look for another location for this. It'd be nice if it was in a class, and not in the global scope.

kienstra replied:
The only concern might be if by chance someone sees this function and starts using it in another plugin. Then, we'd have a harder time deprecating or removing it. 

But I'm not sure how someone would use this function.

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -336,11 +336,7 @@ public function enqueue_block_styles( $name, $type = 'block' ) {
-					"blocks/css/{$type}-{$name}.css",
-					"blocks/{$type}-{$name}.css",
-				)
+				block_lab_template_locations( $name, $type )
```
This edit might not be necessary. It looks to be causing the [failed unit test](https://travis-ci.com/getblocklab/block-lab/jobs/222315202).

lukecarbis replied:
This is addressed in 96550ae.

kienstra replied:
Nice, good idea to create a separate method for stylesheets.

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -434,7 +423,7 @@ public function block_template( $name, $type = 'block' ) {
 				wp_kses_post(
 					// Translators: Placeholder is a file path.
-					sprintf( __( 'Template file %s not found.' ), '<code>' . esc_html( $template_file ) . '</code>' )
+					sprintf( __( 'Template file %s not found.' ), '<code>' . esc_html( $templates[0] ) . '</code>' )
```
`$templates[0]` will probably always be set, unless there's a change to the codebase.

But what do you think about checking that `isset( $templates[0] )`, to future-proof this? Maybe in the conditional above:

```php
if ( ! current_user_can( 'edit_posts' ) || ! isset( $templates[0] ) ) {
	return;
}
```

lukecarbis replied:
Great idea!

lukecarbis replied:
Fixed in 06ad5a0

kienstra replied:
Thanks, that looks great.

kienstra commented on php/post-types/class-block-post.php:
```diff
@@ -325,7 +325,8 @@ public function add_meta_boxes() {
 
 		if ( isset( $post->post_name ) && ! empty( $post->post_name ) ) {
-			$template = block_lab_locate_template( 'blocks/block-' . $post->post_name . '.php', '', true );
+			$locations = block_lab_template_locations( $post->post_name, 'block' );
```
What do you think about not passing the second argument of `'block'`? It looks to be the default argument.

lukecarbis replied:
Fixed in e970caa

kienstra replied:
Thanks, that looks good!

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -398,27 +394,20 @@ public function block_template( $name, $type = 'block' ) {
 		// So lets fix it.
 		if ( empty( $wp_query ) ) {
-			$wp_query = new \WP_Query(); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+			$wp_query = new \WP_Query(); // phpcs: ignore WordPress.WP.GlobalVariablesOverride.Prohibited.
```
Later, it'd be nice to add a unit test for `block_template()`, as that's a really important method. You already added a really nice test for `test_enqueue_block_styles()` 😄 

I don't think this has to happen in this PR, though.

kienstra committed:
Add a simple unit test for block_lab_template_locations()

Test 2 scenarios:
one with the second argument, and one without.

kienstra reviewed (changes requested):
**Looks Good**
*A Few Small Points*


Hi @lukecarbis,
This looks good, and it fixes the issue in my local.

There are just a few small points here. Also, if it's alright, I pushed a small commit to add a unit test.

# Before

When the template was a `block.php` file in a directory with the block name, like `test-image/block.php`, the 'Edit Block' UI didn't show the template as found.

For example, 
`<theme-directory>/blocks/test-image/block.php`

<img width="716" alt="bl-before" src="https://user-images.githubusercontent.com/4063887/62446783-432bd900-b729-11e9-8fa3-a681f0bd9279.png">

# After

As expected, the UI shows this template:

<img width="867" alt="bl-after" src="https://user-images.githubusercontent.com/4063887/62446788-47f08d00-b729-11e9-9c22-ccc303b50c8c.png">

lukecarbis committed:
Remove unused variable

lukecarbis committed:
Add isset check to template locations

lukecarbis committed:
Remove extraneous default variable from function call

lukecarbis committed:
Merge branch 'hotfix/prompt-for-new-template' of https://github.com/getblocklab/block-lab into hotfix/prompt-for-new-template

lukecarbis committed:
Switch preferred template load order

lukecarbis committed:
Add stylesheet template location helper

lukecarbis committed:
Add unit test for block stylesheet locations

lukecarbis commented:
@kienstra Thanks for reviewing! Could you please take a fresh look now?

RobStino reviewed (dismissed):
Tested and this successfully addresses the issue.
Thanks for the efforts on this @lukecarbis and @kienstra

kienstra commented:
Hi @lukecarbis,
Sorry for the delay. I'm working on this now.

kienstra commented:
Hi @lukecarbis, if it's alright, I'm going to push a commit to refactor where utility functions are. It's definitely open for discussion 😄

kienstra committed:
Move 2 helper functions from helpers.php to class-utils.php

This should only be a refactoring,
and shouldn't affect how they work.
But of course this should be tested.

kienstra commented:
Hi @lukecarbis,
What do you think about https://github.com/getblocklab/block-lab/pull/377/commits/a26d3972655d81c7ad28a17b567fcb571df5f43e, which moves this PR's helper functions to a new class, `Utils`?

Like we talked about, another option might be:
1. Put the helper functions, like `block_lab_template_locations()`, in a class like [Loader](https://github.com/getblocklab/block-lab/blob/905af1bc45e709e91f0faa051a4427c887045dfb/php/blocks/class-loader.php)

2. To enable access to the functions, either change [->plugin->components](https://github.com/getblocklab/block-lab/blob/905af1bc45e709e91f0faa051a4427c887045dfb/php/class-plugin-abstract.php#L22) from `protected` to `public`, or add `->loader` as a `public` property of [Plugin](https://github.com/getblocklab/block-lab/blob/905af1bc45e709e91f0faa051a4427c887045dfb/php/class-plugin-abstract.php#L15).

This would work, but it would essentially move [Loader](https://github.com/getblocklab/block-lab/blob/905af1bc45e709e91f0faa051a4427c887045dfb/php/blocks/class-loader.php) to the global scope. So any plugin could call something like `block_lab()->loader->render_block_template()`

As an alternative, this [commit](https://github.com/getblocklab/block-lab/pull/377/commits/a26d3972655d81c7ad28a17b567fcb571df5f43e) creates a new utility function class. 

The class `Plugin` already has utility functions, like [get_assets_url](https://github.com/getblocklab/block-lab/blob/905af1bc45e709e91f0faa051a4427c887045dfb/php/class-plugin-abstract.php#L227). This does something similar, but adds the new functions to `Plugin` via the `->utils` property.

kienstra committed:
Address Travis PHPCS errors

It looks like these were in class-loader.php
So attempt to fix them with the phpcs:ignore commments.

kienstra committed:
Merge branch 'develop' into hotfix/prompt-for-new-template

kienstra commented:
>I think we tend to go with the latter in other parts of the plugin, and it means that we don’t need the class to contain a plugin property.

Ah, interesting. I hadn't thought of that.

lukecarbis commented:
Would it make sense to move the `is_pro()` method to utils?

lukecarbis commented on php/blocks/class-loader.php:
```diff
@@ -394,7 +394,7 @@ public function block_template( $name, $type = 'block' ) {
 		// So lets fix it.
 		if ( empty( $wp_query ) ) {
-			$wp_query = new \WP_Query(); // phpcs: ignore WordPress.WP.GlobalVariablesOverride.Prohibited.
+			$wp_query = new \WP_Query(); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
```
Weird that this wasn't picked up in my version of phpcs. I'll update.

kienstra replied:
All good, I was having a lot of trouble running PHPCS locally.

kienstra committed:
Use Luke's suggestion of a magic method in Plugin

This allows calling methods in Utils
from inside Plugin,
like block_lab()->get_block_lab_template_locations()

kienstra commented:
>If we were fancy, we could even create some sort of pass through magic method to skip the utils property. Like:
block_lab()->get_block_lab_stylesheet_location()

>Would it make sense to move the is_pro() method to utils?

Great ideas, https://github.com/getblocklab/block-lab/pull/377/commits/e74eb0c864172be18edcd6efba1d9be6c3c12d11 applies both of them.

kienstra committed:
Add a test for block_template()

Also, refactor some of the class
so that this deletes all of the files
and directories created.

kienstra committed:
Move block_lab_locate_template() into Utils class

This was a global-scope function before,
and this also enables removing the block_lab_ prefix
of the function.

kienstra committed:
Remove block_lab from functions that were moved into class

Now that these are in the class Utils,
they probably don't need block_lab in
the function name.
They used to be in the global scope.

kienstra commented on php/blocks/class-utils.php:
```diff
@@ -0,0 +1,141 @@
+	 *
+	 * @return string|array
+	 */
+	public function locate_template( $template_names, $path = '', $single = true ) {
```
This is simply moved from `helpers.php`, and not changed at all. It'd be good to have a unit test for this soon.

kienstra replied:
It now has a unit test.

kienstra replied:
Update: there's now a unit test for it.

kienstra committed:
Add DocBlocks for filters in helpers.php

This isn't related to this PR,
but just a general enhancement.

kienstra commented on php/helpers.php:
```diff
@@ -151,82 +151,6 @@ function block_field_config( $name ) {
 	return (array) $block_lab_config->fields[ $name ];
 }
 
-/**
```
Now that there's a class `Utils`, it'd be good to add any new plugin-wide methods there. Unless the function needs to be a plain function for users, like `block_field()`.

kienstra committed:
Add a unit test for locate_template()

Refactor test-class-loader.php,
moving methods into a base class.

kienstra commented:
@lukecarbis,
Good to see you yesterday.

Could you please review this latest changes here when you have a chance? 

I moved more functions into the `Utils` class, and added more tests.

Thanks, Luke!

kienstra committed:
Change the @covers tags, as they referred to the wrong class

These were for magic methods,
but they should refer to their actual class.

kienstra committed:
Rename Base_Template to Abstract_Template

Follow the pattern of using Abstract
in the name of an Abstract class.

lukecarbis reviewed (commented):
These changes look great! Nice work.

Three suggestions:

1. Should `Utils` class be moved to the global namespace? i.e. `Block_Lab\Utils` instead of `Block_Lab\Blocks\Utils`? This would make more sense to me, especially for methods such as `is_pro()`.

2. Should the `block_lab_allowed_svg_tags()` and `block_lab_get_icons()` functions be moved from `helpers.php` to the `Utils` class? I would prefer this, as it would keep `helpers.php` as a space only used for template tags. There's a very small chance it could break backwards compatibility, but it's unlikely anyone would be using these global functions.

3. If we were worried about breaking backwards compatibility with these functions, this could be a good opportunity to include a `deprecated.php`. We could include via `helpers.php` and re-route old functions… e.g.

```PHP
function block_lab_allowed_svg_tags() {
    // Add deprecated notice here.
    block_lab()->allowed_svg_tags();
}
```

This could also be a good place to deprecate actions / filters.

lukecarbis committed:
Minor correction in comment

lukecarbis commented:
Another thought – how about `Utilities` instead of `Utils`? Is `Utils` an accepted convention?

lukecarbis commented:
Going with `Util` to match WordPress.

lukecarbis committed:
Rename Block_Lab\Blocks\Utils to Block_Lab\Util

lukecarbis committed:
Add a way to handle deprecated functions

lukecarbis committed:
Handle deprecated methods in the component abstract

lukecarbis committed:
Move block_lab_get_icons to Util class

lukecarbis committed:
Move block_lab_allowed_svg_tags to Util class

lukecarbis committed:
Move to native WordPress deprecation

lukecarbis commented:
@kienstra Could you please take another look? This hotfix is turning into a rewrite!

lukecarbis committed:
Rename Utils to Util in unit tests

kienstra commented:
>@kienstra Could you please take another look? This hotfix is turning into a rewrite!

Haha, yeah. I'm looking at it now.

kienstra commented:
It looks good so far. I'll continue reviewing in several hours if that's alright.

kienstra commented on php/deprecated.php:
```diff
@@ -0,0 +1,58 @@
+function block_lab_deprecated_function( $function, $version, $replacement ) {
+	_deprecated_function(
+		// filter_var is used for sanitization here as it allows arrow functions ("->").
+		filter_var(
```
What do you think about passing something like `FILTER_SANITIZE_ENCODED` as the 2nd argument of `filter_var()`? 

Good idea to use that function. But so far, it looks like it doesn't sanitize much. 

For example, `filter_var( '<script>doThis();</script>' )` doesn't remove anything.

I'm not sure if `FILTER_SANITIZE_ENCODED` is the right argument, though. It does looks to allow `->`, and it strips `<script>`.

lukecarbis replied:
Great thinking! I went with `FILTER_SANITIZE_STRING` in the end, as:

- It allows arrow functions without encoding them (`->`)
- It removes `<script>` and other malicious things
- We use it elsewhere in the plugin

lukecarbis replied:
Implemented in 7b98dd7.

kienstra replied:
Nice, good idea.

kienstra replied:
Like you mentioned, I also see how passing `FILTER_SANITIZE_STRING` removes `<script>`, but doesn't remove things we need like `->` or `()`

lukecarbis committed:
Add filter type to deprecated function messages

kienstra commented on php/deprecated.php:
```diff
@@ -0,0 +1,59 @@
+ * @param string $version     The version of Block Lab that deprecated the function.
+ * @param string $replacement The function that should have been called.
+ */
+function block_lab_deprecated_function( $function, $version, $replacement ) {
```
Nice helper function!

kienstra commented:
@lukecarbis, is it alright if I push a commit to this to fix a mistake of mine?

lukecarbis commented:
Absolutely

> On 15 Aug 2019, at 10:57 am, Ryan Kienstra <notifications@github.com> wrote:
> 
> @lukecarbis <https://github.com/lukecarbis>, is it alright if I push a commit to this to fix a mistake of mine?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/getblocklab/block-lab/pull/377?email_source=notifications&email_token=AAIL7Q5IZ2END65LF4TDWF3QESSYBA5CNFSM4IJGR4XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4KRBBI#issuecomment-521474181>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAIL7Q5FHXOO4AWJ6OIOCSDQESSYBANCNFSM4IJGR4XA>.
>

kienstra committed:
Like in Luke's commit, throw an Exception if the method doesn't exist

In the __call() method for the Plugin,
I didn't have any handling for the method not existing.
This masked a problem, where Loader
called block_lab->located_template()
before the Utils object was available.
This didn't throw an Exception,
and failed silently.
So this moves the logic that adds Utils to Plugin earlier,
making it available.
It also throws an exception on calling something like
block_lab->nonexistent_method()

lukecarbis commented on php/class-plugin-abstract.php:
```diff
@@ -217,6 +247,14 @@ public function set_version( $file ) {
+	/**
+	 * Instantiates the Util class, with utility methods.
+	 */
+	public function set_util() {
```
Nice move, moving this to the abstract.

kienstra replied:
Thanks!

kienstra committed:
Fix unit tests, which failed with my last commit

This needs to call block_lab->set_util().
Also, this adds a little to test_is_pro().

kienstra commented:
Thanks! Your good idea to throw the Error:

```php
throw new \Error( "Call to undefined method $class::$name()" );
```

...reminded me to add something like that to `Plugin::__call()`. And that revealed that `Loader::retrieve_blocks()` was calling `block_lab->locate_template()` before `block_lab->util` was even set. And it failed silently.

So I added an exception like you added, and set `block_lab->util` earlier.

kienstra commented on block-lab.php:
```diff
@@ -114,7 +117,8 @@ function block_lab() {
 	->set_url( plugin_dir_url( __FILE__ ) )
-	->set_version( __FILE__ );
+	->set_version( __FILE__ )
+	->set_util();
```
This is probably hacky how I set `->util` here. Here's a [little background](https://github.com/getblocklab/block-lab/pull/377#issuecomment-521481994).

Maybe we could eventually abstract all of these methods above, including `set_version()` and `set_util()`, into something like `Plugin::init( $file )` 

But I've probably taken this PR down the rabbit hole enough already 😄

kienstra replied:
Ah, there's a new `Plugin::init()` method in https://github.com/getblocklab/block-lab/pull/269/files#diff-f162295a163d95e718f07b67b57ff1e1R40

lukecarbis replied:
So we should move the `->set_util()` line to the `init` method in the onboarding PR, once this has been merged.

kienstra replied:
Yeah, good point.

kienstra committed:
Add unit tests for get_icons() and allowed_svg_tags()

This is more to prevent regressions,
as these are straight-forward functions.

lukecarbis commented:
✅ Looks good to me.

Need your review to merge.

kienstra reviewed (approved):
**Approved**

Hi @lukecarbis,
Thanks for your great work and patience. This looks good!

kienstra commented:
Like with any big refactoring PR, the plugin could probably use a full regression test after this.
